In [112]:
import pandas as pd
import numpy as np

# reads dataset and replaces linebreaks and missing values with ''
data = []
with open('../datasets/chronic_kidney_disease_full.arff', "r") as f:
    for line in f:
        line = line.replace('\n', '')
        line = line.replace('?', '')
        data.append(line.split(','))

# gives names to the columns
names = ['age', 'bp', 'sg', 'al', 'su', 'rbc', 'pc', 'pcc', 'ba',
         'bgr', 'bu',  'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc',
         'rbcc', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane',
         'class', 'no_name']
    
df = pd.DataFrame(data[145:], columns=names)
# np.nan is recognised as missing value for Pandas functions e.g. isnull()
df = df.replace('', np.nan)

df

,age,bp,sg,al,su,rbc,pc,pcc,ba,bgr,...,wbcc,rbcc,htn,dm,cad,appet,pe,ane,class,no_name
0,48,80,1.020,1,0,NaN,normal,notpresent,notpresent,121,...,7800,5.2,yes,yes,no,good,no,no,ckd,None
1,7,50,1.020,4,0,NaN,normal,notpresent,notpresent,NaN,...,6000,NaN,no,no,no,good,no,no,ckd,None
2,62,80,1.010,2,3,normal,normal,notpresent,notpresent,423,...,7500,NaN,no,yes,no,poor,no,yes,ckd,None
3,48,70,1.005,4,0,normal,abnormal,present,notpresent,117,...,6700,3.9,yes,no,no,poor,yes,yes,ckd,None
4,51,80,1.010,2,0,normal,normal,notpresent,notpresent,106,...,7300,4.6,no,no,no,good,no,no,ckd,None
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
397,12,80,1.020,0,0,normal,normal,notpresent,notpresent,100,...,6600,5.4,no,no,no,good,no,no,notckd,None
398,17,60,1.025,0,0,normal,normal,notpresent,notpresent,114,...,7200,5.9,no,no,no,good,no,no,notckd,None
399,58,80,1.025,0,0,normal,normal,notpresent,notpresent,131,...,6800,6.1,no,no,no,good,no,no,notckd,None
400,NaN,None,None,None,None,None,None,None,None,None,...,None,None,None,None,None,None,None,None,None,None


In [113]:
# Drop unwanted columns
df.drop(columns=['rbc', 'pc','pcc','ba','sc','htn','dm','cad','appet','pe','ane', 'no_name'], inplace=True)
df

,age,bp,sg,al,su,bgr,bu,sod,pot,hemo,pcv,wbcc,rbcc,class
0,48,80,1.020,1,0,121,36,NaN,NaN,15.4,44,7800,5.2,ckd
1,7,50,1.020,4,0,NaN,18,NaN,NaN,11.3,38,6000,NaN,ckd
2,62,80,1.010,2,3,423,53,NaN,NaN,9.6,31,7500,NaN,ckd
3,48,70,1.005,4,0,117,56,111,2.5,11.2,32,6700,3.9,ckd
4,51,80,1.010,2,0,106,26,NaN,NaN,11.6,35,7300,4.6,ckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
397,12,80,1.020,0,0,100,26,137,4.4,15.8,49,6600,5.4,notckd
398,17,60,1.025,0,0,114,50,135,4.9,14.2,51,7200,5.9,notckd
399,58,80,1.025,0,0,131,18,141,3.5,15.8,53,6800,6.1,notckd
400,NaN,None,None,None,None,None,None,None,None,None,None,None,None,None


In [114]:
# converts hemoglobin values from g/dl to d/l by multiplying by 10
# original type (str) is first converted to float, multiplied and then converted back
df['hemo'] = df['hemo'].astype('Float64')
df['hemo'] = df['hemo'].apply(lambda x: x*10)
df['hemo'] = df['hemo'].astype(str)
df

,age,bp,sg,al,su,bgr,bu,sod,pot,hemo,pcv,wbcc,rbcc,class
0,48,80,1.020,1,0,121,36,NaN,NaN,154.0,44,7800,5.2,ckd
1,7,50,1.020,4,0,NaN,18,NaN,NaN,113.0,38,6000,NaN,ckd
2,62,80,1.010,2,3,423,53,NaN,NaN,96.0,31,7500,NaN,ckd
3,48,70,1.005,4,0,117,56,111,2.5,112.0,32,6700,3.9,ckd
4,51,80,1.010,2,0,106,26,NaN,NaN,116.0,35,7300,4.6,ckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
397,12,80,1.020,0,0,100,26,137,4.4,158.0,49,6600,5.4,notckd
398,17,60,1.025,0,0,114,50,135,4.9,142.0,51,7200,5.9,notckd
399,58,80,1.025,0,0,131,18,141,3.5,158.0,53,6800,6.1,notckd
400,NaN,None,None,None,None,None,None,None,None,<NA>,None,None,None,None


In [115]:
# recodes class values 'ckd' and 'notckd' (chronic kidney disease, control) to 'a' and 'c'
df['class'] = df['class'].replace({'ckd':'a', 'notckd':'c'})
df

,age,bp,sg,al,su,bgr,bu,sod,pot,hemo,pcv,wbcc,rbcc,class
0,48,80,1.020,1,0,121,36,NaN,NaN,154.0,44,7800,5.2,a
1,7,50,1.020,4,0,NaN,18,NaN,NaN,113.0,38,6000,NaN,a
2,62,80,1.010,2,3,423,53,NaN,NaN,96.0,31,7500,NaN,a
3,48,70,1.005,4,0,117,56,111,2.5,112.0,32,6700,3.9,a
4,51,80,1.010,2,0,106,26,NaN,NaN,116.0,35,7300,4.6,a
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
397,12,80,1.020,0,0,100,26,137,4.4,158.0,49,6600,5.4,c
398,17,60,1.025,0,0,114,50,135,4.9,142.0,51,7200,5.9,c
399,58,80,1.025,0,0,131,18,141,3.5,158.0,53,6800,6.1,c
400,NaN,None,None,None,None,None,None,None,None,<NA>,None,None,None,None


In [116]:
# remove rows with three or more missing values
df = df[df.isnull().sum(axis=1) < 3]
df

,age,bp,sg,al,su,bgr,bu,sod,pot,hemo,pcv,wbcc,rbcc,class
0,48,80,1.020,1,0,121,36,NaN,NaN,154.0,44,7800,5.2,a
3,48,70,1.005,4,0,117,56,111,2.5,112.0,32,6700,3.9,a
4,51,80,1.010,2,0,106,26,NaN,NaN,116.0,35,7300,4.6,a
5,60,90,1.015,3,0,74,25,142,3.2,122.0,39,7800,4.4,a
6,68,70,1.010,0,0,100,54,104,4.0,124.0,36,NaN,NaN,a
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
395,55,80,1.020,0,0,140,49,150,4.9,157.0,47,6700,4.9,c
396,42,70,1.025,0,0,75,31,141,3.5,165.0,54,7800,6.2,c
397,12,80,1.020,0,0,100,26,137,4.4,158.0,49,6600,5.4,c
398,17,60,1.025,0,0,114,50,135,4.9,142.0,51,7200,5.9,c


In [117]:
# reindex dataframe to correctly show the number of rows left
df = df.reset_index(drop=True)
df

,age,bp,sg,al,su,bgr,bu,sod,pot,hemo,pcv,wbcc,rbcc,class
0,48,80,1.020,1,0,121,36,NaN,NaN,154.0,44,7800,5.2,a
1,48,70,1.005,4,0,117,56,111,2.5,112.0,32,6700,3.9,a
2,51,80,1.010,2,0,106,26,NaN,NaN,116.0,35,7300,4.6,a
3,60,90,1.015,3,0,74,25,142,3.2,122.0,39,7800,4.4,a
4,68,70,1.010,0,0,100,54,104,4.0,124.0,36,NaN,NaN,a
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
264,55,80,1.020,0,0,140,49,150,4.9,157.0,47,6700,4.9,c
265,42,70,1.025,0,0,75,31,141,3.5,165.0,54,7800,6.2,c
266,12,80,1.020,0,0,100,26,137,4.4,158.0,49,6600,5.4,c
267,17,60,1.025,0,0,114,50,135,4.9,142.0,51,7200,5.9,c


---

<h3>Task 2</h3>
<i>Split the data frame into two data frames, one for the affected individuals and one for the control individuals. Display the data frames, and indicate the number of rows in each data frame.</i>

---

First, we create a new data frame with only affected individuals based on the old data frame. We know that <b>all affected individuals have letter "a" in the class column</b>. So creating a new data frame with only affected individuals goes as followed...

In [119]:
df_affected = df[df["class"] == "a"] #create a new data frame with filter that has only people with "a" in the class.
df_affected #display the new data generated.

,age,bp,sg,al,su,bgr,bu,sod,pot,hemo,pcv,wbcc,rbcc,class
0,48,80,1.020,1,0,121,36,NaN,NaN,154.0,44,7800,5.2,a
1,48,70,1.005,4,0,117,56,111,2.5,112.0,32,6700,3.9,a
2,51,80,1.010,2,0,106,26,NaN,NaN,116.0,35,7300,4.6,a
3,60,90,1.015,3,0,74,25,142,3.2,122.0,39,7800,4.4,a
4,68,70,1.010,0,0,100,54,104,4.0,124.0,36,NaN,NaN,a
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
124,62,90,1.020,2,1,169,48,138,2.9,134.0,47,11000,6.1,a
125,64,90,1.015,3,2,463,64,135,4.1,122.0,40,9800,4.6,a
126,48,110,1.015,3,0,106,215,120,5.7,86.0,26,5000,2.5,a
127,59,70,1.010,1,3,424,55,138,4.5,126.0,37,10200,4.1,a


<b>Result:</b> Here we can see that we have 127 unique individuals affected by the disease. We created a new data frame called "df_affected", which contains only the people affected with the Chronic Kidney Disease.

---
So we continue, we have to now use the same technique to separate the people that are controlled for the individuals who's disease is progressing.

In [121]:
df_controlled = df[df["class"] == "c"] #create a new data frame with filter that has only people with "c" in the class.
df_controlled #display the new data generated.

,age,bp,sg,al,su,bgr,bu,sod,pot,hemo,pcv,wbcc,rbcc,class
129,40,80,1.025,0,0,140,10,135,5.0,150.0,48,10400,4.5,c
130,23,80,1.025,0,0,70,36,150,4.6,170.0,52,9800,5.0,c
131,45,80,1.025,0,0,82,49,147,4.4,159.0,46,9100,4.7,c
132,57,80,1.025,0,0,119,17,135,4.7,154.0,42,6200,6.2,c
133,51,60,1.025,0,0,99,38,135,3.7,130.0,49,8300,5.2,c
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
264,55,80,1.020,0,0,140,49,150,4.9,157.0,47,6700,4.9,c
265,42,70,1.025,0,0,75,31,141,3.5,165.0,54,7800,6.2,c
266,12,80,1.020,0,0,100,26,137,4.4,158.0,49,6600,5.4,c
267,17,60,1.025,0,0,114,50,135,4.9,142.0,51,7200,5.9,c


<b>Result:</b>  Here we can see the list and amount of the contolled individuas, which is 138. Now we have to different data frames that have their on patients listed correctly.

<i>df_affected</i> = All the patients with the disease. (128 patients.) </br>
<i>df_controlled</i> = All the patients with sign of disease and being supervised. ( 139 patients. )

---

<h3>Task 3</h3>
For each data frame, calculate the basic statistics for each column, and provide clear, readable histograms for each numerical column. Do you see any outliers? If so, how would you handle them?


<h3>Task 4</h3>
Calculate the correlation matrix and visualize it for each data frame. Clearly describe the results and your interpretation for it.
